File: sdk/packages/python/iii/src/iii/baggage_span_processor.py
"""Baggage -> span attribute processor."""

from __future__ import annotations

from typing import Sequence

from opentelemetry import baggage
from opentelemetry.context import Context
from opentelemetry.sdk.trace import ReadableSpan, Span, SpanProcessor

#: DEFAULT_ALLOWLIST drift across languages would break worker chains;
#: lockstep tests in each SDK pin this constant at CI time.
DEFAULT_ALLOWLIST: tuple[str, ...] = (
    "iii.session.id",
    "iii.message.id",
    "iii.function.id",
)


class BaggageSpanProcessor(SpanProcessor):

    def __init__(self, allowlist: Sequence[str] = DEFAULT_ALLOWLIST) -> None:
        self._allowlist: tuple[str, ...] = tuple(allowlist)

    def on_start(self, span: Span, parent_context: Context | None = None) -> None:
        # NoOp guard: skip allocation when sampler drops the span.
        if not span.is_recording():
            return

        for key in self._allowlist:
            value = baggage.get_baggage(key, parent_context)
            if value is not None:
                span.set_attribute(key, str(value))

    def on_end(self, span: ReadableSpan) -> None:  # noqa: ARG002
        pass

    def shutdown(self) -> None:
        pass

    def force_flush(self, timeout_millis: int = 30000) -> bool:  # noqa: ARG002
        return True

File: sdk/packages/node/iii/src/telemetry-system/baggage-span-processor.ts
// Baggage -> span attribute processor.

import type { Context } from '@opentelemetry/api'
import { propagation } from '@opentelemetry/api'
import type { ReadableSpan, Span, SpanProcessor } from '@opentelemetry/sdk-trace-base'

/** DEFAULT_ALLOWLIST drift across languages would break worker chains;
 * lockstep tests in each SDK pin this constant at CI time. */
export const DEFAULT_ALLOWLIST: readonly string[] = [
  'iii.session.id',
  'iii.message.id',
  'iii.function.id',
] as const

export class BaggageSpanProcessor implements SpanProcessor {
  private readonly allowlist: readonly string[]

  constructor(allowlist: readonly string[] = DEFAULT_ALLOWLIST) {
    this.allowlist = allowlist
  }

  onStart(span: Span, parentContext: Context): void {
    // NoOp guard: skip allocation when sampler drops the span.
    if (!span.isRecording()) {
      return
    }

    const baggage = propagation.getBaggage(parentContext)
    if (!baggage) {
      return
    }

    for (const key of this.allowlist) {
      const entry = baggage.getEntry(key)
      if (entry) {
        span.setAttribute(key, entry.value)
      }
    }
  }

  onEnd(_span: ReadableSpan): void {
    /* no-op */
  }

  async shutdown(): Promise<void> {
    /* no-op */
  }

  async forceFlush(): Promise<void> {
    /* no-op */
  }
}
